components/navigable-container/index.js

```diff
@@ -0,0 +1,167 @@
+/**
+ * External Dependencies
+ */
+import { omit, noop } from 'lodash';
+
+/**
+ * WordPress Dependencies
+ */
+import { Component } from '@wordpress/element';
+import { focus, keycodes } from '@wordpress/utils';
+
+/**
+ * Module Constants
+ */
+const { UP, DOWN, LEFT, RIGHT, TAB } = keycodes;
+
+function cycleValue( value, total, offset ) {
+	const nextValue = value + offset;
+	if ( nextValue < 0 ) {
+		return total + nextValue;
+	} else if ( nextValue >= total ) {
+		return nextValue - total;
+	}
+
+	return nextValue;
+}
+
+class NavigableContainer extends Component {
+	constructor() {
+		super( ...arguments );
+		this.bindContainer = this.bindContainer.bind( this );
+		this.onKeyDown = this.onKeyDown.bind( this );
+
+		this.getFocusableContext = this.getFocusableContext.bind( this );
+		this.getFocusableIndex = this.getFocusableIndex.bind( this );
+	}
+
+	bindContainer( ref ) {
+		const { handleRef = noop } = this.props;
+		this.container = ref;
+		handleRef( ref );
+	}
+
+	getFocusableContext( target ) {
+		const { deep, onlyBrowserTabstops } = this.props;
+		const finder = onlyBrowserTabstops ? focus.tabbable : focus.focusable;
+		const focusables = finder
+			.find( this.container )
+			.filter( ( node ) => deep || node.parentElement === this.container );
+
+		const index = this.getFocusableIndex( focusables, target );
+		if ( index > -1 && target ) {
+			return { index, target, focusables };
+		}
+		return null;
+	}
+
+	getFocusableIndex( focusables, target ) {
+		const directIndex = focusables.indexOf( target );
+		if ( directIndex !== -1 ) {
+			return directIndex;
+		}
+	}
+
+	onKeyDown( event ) {
+		if ( this.props.onKeyDown ) {
+			this.props.onKeyDown( event );
+		}
+
+		const { keyCode } = event;
+		const { getFocusableContext } = this;
+		const { cycle = true, eventToOffset, onNavigate = noop, preventBubblingArrowEvents = true, preventBubblingTabEvents = false } = this.props;
+
+		const offset = eventToOffset( event );
+
+		const stopNavigationKey = ( preventBubblingArrowEvents && [ LEFT, RIGHT, DOWN, UP ].indexOf( keyCode ) > -1 ) ||
```

Comment:
Could this be simplified mike `preventBubblingArrowEvents && offset !== 0`

Reply:
Oh I see, it's relying on `preventBubblingTabEvents` or `preventBubblingArrowEvents` which make things a bit different. Can't we use the same prop? `preventBubblingNavigationEvents`?

Reply:
They can't use the same prop. Think of the Inserter Example


* TabbingCycle
  * Input
  * TabList (cares about arrows)
    * tab1
    * tab2
  * TabView (cares about arrows)
    * item1
    * item2

If the TabList (which needs to stop arrow keys bubbling also stopped tab keys bubbling), then the TabbingCycle wouldn't work. The TabbingCycle needs to receive the tab key, otherwise it can't force tab to cycle within itself. They need to be separate. Does that make sense?

Reply:
I've had to rename the prop because we need to `preventDefault` in many situations as well, not just `stopPropagation`.